File: 백준/Silver/9012. 괄호/괄호.js
const fs = require("fs");
const input = fs.readFileSync('/dev/stdin').toString().split("\n");

const T = Number(input[0]);

let answer = ""

for (let i = 1; i <= T; i++) {
  let str = input[i];
  let stack = [];

  for (let char of str) {
    if (char === "(") {
      stack.push(char)
    } else if (char === ")") {
      if (stack.length > 0 && stack[stack.length - 1] === "(") {
        stack.pop()
      } else {
        stack.push(char)
      }
    }
  }
  if (stack.length > 0) {
    answer += "NO\n"
  } else {
    answer += "YES\n"
  }
}

console.log(answer)

File: 백준/Silver/9012. 괄호/괄호.py
import sys
from collections import deque

N = int(sys.stdin.readline())
result = []
for _ in range(N):
  stack = deque()
  str = sys.stdin.readline().rstrip()

  for char in str:
    if char == "(":
      stack.append(char)
    elif char == ")":
      if stack and stack[-1] == "(":
        stack.pop()
      else:
        stack.append(char)

  
  if stack:
    result.append("NO")
  else:
    result.append("YES")

print("\n".join(result))
  